drivers/sensor/sensor_shell.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2018 Diego Sueiro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <shell/shell.h>
+#include <stdlib.h>
+#include <string.h>
+#include <ctype.h>
+#include <device.h>
+#include <sensor.h>
+#include <syscall_handler.h>
+
+extern struct device __device_init_start[];
+extern struct device __device_init_end[];
+
+const char *sensor_channel_name[SENSOR_CHAN_ALL] = {
+	[SENSOR_CHAN_ACCEL_X] =		"accel_x",
+	[SENSOR_CHAN_ACCEL_Y] =		"accel_y",
+	[SENSOR_CHAN_ACCEL_Z] =		"accel_z",
+	[SENSOR_CHAN_ACCEL_XYZ] =	"accel_xyz",
+	[SENSOR_CHAN_GYRO_X] =		"gyro_x",
+	[SENSOR_CHAN_GYRO_Y] =		"gyro_y",
+	[SENSOR_CHAN_GYRO_Z] =		"gyro_z",
+	[SENSOR_CHAN_GYRO_XYZ] =	"gyro_xyz",
+	[SENSOR_CHAN_MAGN_X] =		"magn_x",
+	[SENSOR_CHAN_MAGN_Y] =		"magn_y",
+	[SENSOR_CHAN_MAGN_Z] =		"magn_z",
+	[SENSOR_CHAN_MAGN_XYZ] =	"magn_xyz",
+	[SENSOR_CHAN_DIE_TEMP] =	"die_temp",
+	[SENSOR_CHAN_AMBIENT_TEMP] =	"ambient_temp",
+	[SENSOR_CHAN_PRESS] =		"press",
+	[SENSOR_CHAN_PROX] =		"prox",
+	[SENSOR_CHAN_HUMIDITY] =	"humidity",
+	[SENSOR_CHAN_LIGHT] =		"light",
+	[SENSOR_CHAN_IR] =		"ir",
+	[SENSOR_CHAN_RED] =		"red",
+	[SENSOR_CHAN_GREEN] =		"green",
+	[SENSOR_CHAN_BLUE] =		"blue",
+	[SENSOR_CHAN_ALTITUDE] =	"altitude",
+	[SENSOR_CHAN_PM_1_0] =		"pm_1_0",
+	[SENSOR_CHAN_PM_2_5] =		"pm_2_5",
+	[SENSOR_CHAN_PM_10] =		"pm_10",
+	[SENSOR_CHAN_DISTANCE] =	"distance",
+	[SENSOR_CHAN_CO2] =		"co2",
+	[SENSOR_CHAN_VOC] =		"voc",
+	[SENSOR_CHAN_VOLTAGE] =		"voltage",
+	[SENSOR_CHAN_CURRENT] =		"current",
+	[SENSOR_CHAN_ROTATION] =	"rotation",
+};
+
+static int cmd_list_sensor_channels(const struct shell *shell,
+				   size_t argc, char *argv[])
+{
+	int err;
+	struct device *dev;
+	struct sensor_value value[3];
+
+	err = shell_cmd_precheck(shell, (argc >= 2), NULL, 0);
+	if (err) {
+		return err;
+	}
+
+	dev = device_get_binding(argv[1]);
+
+	if (dev != NULL) {
+		k_thread_access_grant(k_current_get(), dev, NULL);
+		if (Z_SYSCALL_DRIVER_SENSOR(dev, sample_fetch)) {
```

Comment:
not a public API

Reply:
I think I understand the intent here, but this isn't the right way to go about it.

> Even having the CONFIG_USERSAPCE=y set when calling sensor_sample_fetch it seems that it is calling the _impl_sensor_sample_fetch directly instead of Z_SYSCALL_HANDLER(sensor_sample_fetch, dev) defined in the drivers/sensor/sensor_handlers.c file

Yes, that's correct, handler checks are only performed when called from user mode, not supervisor mode.

> Do you have another suggestion to validade if a device is a sensor

The solution would be to have the shell subsystem shell() thread run in user mode.
Right now, everything is running in supervisor mode which assumes that all inputs are sane, there's  at best incomplete checking done. This is really a problem with the shell subsystem. There are all kinds of ways you can hose the system with bad shell commands, not just incorrect driver names. As a debugging/demo/test subsystem, historically people have been fine with the shell behaving in this way.

I'd simply drop the validation you are trying to do here; in order to properly have a shell which is robust against bad input, the whole thing needs to be in user mode. That way, if you start making API calls with bad input, the existing handlers will catch it.

Reply:
@andrewboie ,

Thanks for the explanation. I removed the checks and the dependency on USERSPACE.
But I really think that we should have the shell running in a more secure environment with extra API call checks.

If you have time, please test this shell module by calling the following command: `sensor get UART_0` and `sensor get I2C_0`. I don't think this is is an acceptable behavior.

I tried to find in the shell subsys where it is dispatching the commands as threads but I couldn't find. Maybe we make use of the `k_thread_user_mode_enter`. But I'm not sure how complex can it be.

Reply:
> But I really think that we should have the shell running in a more secure environment with extra API call checks.

I understand where you are coming from but I think the only way to really assert security of the shell is to run it in user mode. Imagine if the Linux kernel console had an interactive command prompt that ran in kernel context...

>  I don't think this is is an acceptable behavior.

It depends. If all we are using shell for is debugging or demonstrations, then we don't really care about the security. If this is an end user facing feature, then it would be insane to expose it in its current form, and not just for being able to pass in bad driver names.

From my point of view, guarding against bad driver names for this specific command without actually running the entire shell in a reduced CPU privilege level is like locking the door and leaving all the windows open. There are so many shell commands implemented, any one of them could be an attack surface.

> I tried to find in the shell subsys where it is dispatching the commands as threads

Given that the shell subsys is processing arbitrary user input, I think it would only do to have the main shell thread started in shell.c start with K_USER options. It won't be a simple as that, there will be a lot more work to do since the shell subsystem was written before user mode was a part of Zephyr.

Reply:
@andrewboie ,

I was thinking about creating an app that interacts with the user by printing and receiving data from UART. Having the shell subsys seems to be perfect for that. Lots of devices in the industrial field only have a serial port to interact with the user.

Do you think we should file an issue to get the shell subsys running in user mode and see what the others developers say?